cli/command/stack/kubernetes/list.go

```diff
@@ -43,6 +52,58 @@ func getStacks(kubeCli *KubeCli, opts options.List) ([]*formatter.Stack, error)
 	}
 	return formattedStacks, nil
 }
+func getStacksWithAllNamespaces(dockerCli command.Cli, kubeCli *KubeCli, opts options.List) ([]*formatter.Stack, error) {
+	stacks, err := getStacks(kubeCli, opts)
+	if err == nil || !apierrs.IsForbidden(err) {
+		return stacks, err
+	}
+	nms, err := getUserVisibleNamespaces(dockerCli, err)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to query user visible namespaces")
+	}
+	opts.AllNamespaces = false
+	for _, namespace := range nms.Items {
+		kubeCli.kubeNamespace = namespace.Name
+		ss, err := getStacks(kubeCli, opts)
+		if err != nil {
+			return nil, err
+		}
+		stacks = append(stacks, ss...)
+	}
+	return stacks, nil
+}
+
+func getUserVisibleNamespaces(dockerCli command.Cli, previousErr error) (*core_v1.NamespaceList, error) {
```

Comment:
@thaJeztah @mat007 I think it's a little weird to pass in `previousErr` to this function so that we can return it in certain cases.

@thaJeztah, what did you have in mind here?

Reply:
I suppose I could either return also a boolean meaning «404», just ignore the call? Or create a custom error to be returned just for the sake a filtering right after the call?
The `previousErr` remains the solution with the less extra code, which is why I went for it.

Reply:
I think your previous solution (return an error and then have `getStacksWithAllNamespaces` check for that err and return the previous error) was actually better, which is why I'm wondering if @thaJeztah had something different in mind.

Reply:
My suggestion (here https://github.com/docker/cli/pull/1034#discussion_r188932239) was because the previous solution was discarding all errors returned by UCP, which was definitely not the right thing to do.

I see I made one mistake in that example, which probably lead to @mat007 rewriting it (updated version below);

<details>

<summary>updated example:</summary>

```go
func getStacksWithAllNamespaces(dockerCli command.Cli, kubeCli *KubeCli, opts options.List) ([]*formatter.Stack, error) {
	stacks, err := getStacks(kubeCli, opts)
	if err == nil || !apierrs.IsForbidden(err) {
		return stacks, err
	}
	nms, err2 := getUserVisibleNamespaces(dockerCli)
	if err2 != nil {
		return nil, errors.Wrap(err2, "Failed to query user visible namespaces")
	}
	if nms == nil {
		return nil, errors.Wrap(err, "Failed to retrieve stacks")
	}
	opts.AllNamespaces = false
	for _, namespace := range nms.Items {
		kubeCli.kubeNamespace = namespace.Name
		ss, err := getStacks(kubeCli, opts)
		if err != nil {
			return nil, err
		}
		stacks = append(stacks, ss...)
	}
	return stacks, nil
}

func getUserVisibleNamespaces(dockerCli command.Cli) (*core_v1.NamespaceList, error) {
	host := dockerCli.Client().DaemonHost()
	endpoint, err := url.Parse(host)
	if err != nil {
		return nil, err
	}
	endpoint.Scheme = "https"
	endpoint.Path = "/kubernetesNamespaces"
	resp, err := dockerCli.Client().HTTPClient().Get(endpoint.String())
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()
	body, err := ioutil.ReadAll(resp.Body)
	if err != nil {
		return nil, errors.Wrapf(err, "received %d status and unable to read response", resp.StatusCode)
	}
	switch resp.StatusCode {
	case http.StatusOK:
	  namespaces := &core_v1.NamespaceList{}
		if err := json.Unmarshal(body, namespaces); err != nil {
			return nil, errors.Wrapf(err, "unmarshal failed: %s", string(body))
		}
	  return namespaces, nil
	case http.StatusNotFound:
		// UCP API not present
		return nil, nil
	default:
		return nil, fmt.Errorf("received %d status while retrieving namespaces: %s", resp.StatusCode, string(body))
	}
}
```

</details>


<p><br /></p>
An alternative could be to flip the order (dummy code);

```go
if stacks, err := getUCPStacks(dockerCli, opts); err != nil || stacks != nil {
    return stacks, err
}
return getStacks(kubeCli, opts)
```

In the above;

- In case of a 404, `getUCPStacks()` returns no error, and `nil` for stacks, in that case, fall back to plain kubernetes
- Any other error is an issue with the UCP endpoint, so should be handled. In that case, don't fallback to plain kubernetes


- Advantage: There's no back and forth between trying UCP and k8s; currently, we try k8s, get an error, try UCP, discover it's not there, and head back to the original k8s error
- Disadvantage: It's odd for Docker CE clients to try an API that won't be there for 99% of the users

Reply:
I believe we cannot swap the function calls as what we want is first Kubernetes, then UCP only if Kubernetes fails for lack of user rights.

@wsong we have 3 possible results when querying UCP : not there (404), there but error'ed and valid result, so we need to distinguish between the three
@thaJeztah ah yes, I suppose we could test for `nil` to distinguish a 404 from `[]string{}` an empty result, OK I'll do that instead

Thank you both!